File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        PriorityQueue<Pair<Integer, Integer>> maxHeap = new PriorityQueue<>((a, b) ->
        a.getValue() - b.getValue());
        int[] anw = new int[k];

        for (int num : nums) map.put(num, map.getOrDefault(num, 0) + 1);

        for (Map.Entry<Integer, Integer> it : map.entrySet()) {
            maxHeap.add(new Pair<>(it.getKey(), it.getValue()));
            if (maxHeap.size() > k) maxHeap.poll();
        }

        for (int i = 0; i < k; i++) {
            anw[i] = (maxHeap.poll().getKey());
        }
        return anw;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;

        for (int num : nums) map[num]++;


        auto cmp = [] (const pair<int, int>& a, const pair<int, int>& b) {
            return a.second > b.second;
        };

        priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> pq;

        for (auto [key, value] : map) {
            pq.emplace(pair<int, int>(key, value));

            if (pq.size() > k) pq.pop();
        }

        vector<int> res;

        while (!pq.empty()) {
            res.push_back(pq.top().first);
            pq.pop();
        }

        return res;
    }
};